### DISCUSSION TITLE
Asynchronously call spdlog::get()

### DISCUSSION THREAD

**Discussion by gk017** [gabime/spdlog#2189] [Q&A] [answered]
I just found out that asynchronous calls of spdlog::get() seems not to work as I expected. Am I doing something wrong?
I have created a minimal example with a test class that uses the logger and a google test that uses the test class.

This is my test class:
```
class SpdLogTestClass
{
public:
  SpdLogTestClass()
  {
      m_log = spdlog::get("Test");
      if (!m_log)
      {
          m_log = spdlog::stdout_color_mt("Test");
      }
  }

  [[nodiscard]] UUID id()
  {
      return m_id;
  }
private:
  UUID m_id;
  std::shared_ptr<spdlog::logger> m_log;
};
```
and this is the code in my test that uses the test class:
```
std::function<UUID(void)> lambda = []()
{
    SpdLogTestClass ltc;
    return ltc.id();
};
std::vector<std::future<UUID>> futures;
for(int i = 0; i < 3; i++)
    futures.emplace_back(std::async(std::launch::async, lambda));
for(auto & f : futures)
    std::cout << f.get() << std::endl;
```
Not always but often I catch this exception and execution of test stops before I see all three expected lines of output:
```
logger with name 'Test' already exists
```

So far I have tried to create the test class instances synchronously and it works as expected: for the first instance the logger is created and all following instances get a pointer to the already existing logger.

Has someone an idea what I am doing wrong?

**Comment by tt4g** [comment]
The variable name is wrong.
Assign the logger you created to `m_log`.

```diff
   SpdLogTestClass()
   {
       m_log = spdlog::get("Test");
       if (!m_log)
       {
-          m_logger = spdlog::stdout_color_mt("Test");
+          m_logger = spdlog::stdout_color_mt("Test");
       }
   }
```

**Reply by gk017** [reply]
Ah, yes, but this was because of simplifying for posting here. I edited the original question.

**Comment by gabime** [comment, marked as answer]
seems you need to mutex protect the check+logger creation block.

**Reply by gk017** [reply]
Thanks for the hint! Never had to use a mutex before but that solves the problem:

```
std::mutex m;

class SpdLogTestClass
{
public:
  SpdLogTestClass()
  {
      std::lock_guard<std::mutex> guard(m);
      m_log = spdlog::get("Test");
      if (!m_log)
      {
          m_log = spdlog::stdout_color_mt("Test");
      }
  }

  [[nodiscard]] UUID id()
  {
      return m_id;
  }
private:
  UUID m_id;
  std::shared_ptr<spdlog::logger> m_log;
};
```
